### ISSUE TITLE
"nixos-rebuild switch" kills contaner networking

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#21745]
## Issue description

```nixos-rebuild switch``` often (but not always) executes ```systemctl stop network-setup``` then ```systemctl start network-setup```.

It kills ```nixos-container```'s network.

### Steps to reproduce

test.nix:
```
with import <nixpkgs/nixos/lib/testing.nix> { system = builtins.currentSystem; };

(makeTest { 

  nodes = {

    client = { lib, pkgs, ... }: {
      networking.firewall.enable = false;
      services.xserver.enable = true;

      services.xserver.displayManager.auto.enable = true;

      services.xserver.windowManager.default = "icewm";
      services.xserver.windowManager.icewm.enable = true;
      services.xserver.desktopManager.default = "none";

      networking.bridges.br0.interfaces = [ "eth1" ];
      networking.interfaces = {
        eth1.ip4 = lib.mkOverride 0 [ ];
        br0.ip4 = [{ address = "192.168.1.1"; prefixLength = 24; }];
      };

      containers.webserver = {
        autoStart = true;
        privateNetwork = true;
        hostBridge = "br0";
        localAddress = "192.168.1.222/24";
        config = { 
          networking.firewall.allowPing = true;
        };
      };

    };
  };

  testScript = ''
    startAll;

    $client->waitForX;

    $client->sleep(1);
    $client->execute("xterm &");
    $client->sleep(1);
    $client->sendChars("ping 192.168.1.222\n");

    $client->sleep(1);
    $client->execute("xterm &");
    $client->sleep(1);
    $client->sendChars("nixos-container root-login webserver\n");
    $client->sendChars("ping 192.168.1.1\n");

    $client->sleep(5);
    $client->execute("xterm &");
    $client->sleep(1);
    $client->sendChars("systemctl stop network-setup\n");
    $client->sleep(1);
    $client->sendChars("systemctl start network-setup\n");

    $client->sleep(10);
    $client->screenshot("screen");

    $client->sleep(10000); # allow interaction
  '';

}).driver
```

run as: (https://github.com/NixOS/nixpkgs/pull/21697 may be required for ```sendChars``` to send '.' and '-')
```
tests='eval $ENV{testScript}' $(nix-build test.nix)/bin/nixos-test-driver
```

![screen](https://cloud.githubusercontent.com/assets/508305/21746513/a235a38c-d53f-11e6-8035-7aa7c9f039b5.png)

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
17.03.git.7215118 (Gorilla)

* Nix version: (run `nix-env --version`)
nix-env (Nix) 1.11.5

* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
"17.03.git.7215118c1b"

**Comment by fpletz** [comment]
The problem is that scripted networking might recreate the bridge if something changes. Other virtualization options like libvirt suffer from the same problem if bridges managed by NixOS are used. Using Using systemd's networkd fixes this problem for me.

The scripted networking should be fixed somehow, though the solution might be a bit tricky. When we enslave an interface to a bridge through the NixOS config and switch to another config without that interface on that bridge, should we remove it from the bridge? I'm not sure if we can detect that on action and but recreating the bridge solves that - and removes all dynamically enslaved interfaces.

We have two possible solutions IMHO:

  1. We don't recreate bridges. Interfaces will not be removed from bridges on system switch/activation.
  2. Ensure all services depending on the bridge are restarted. In case of NixOS containers, we can do it, but other software will still break.

I would go for the first option.

**Comment by volth** [comment]
What if ```nixos-rebuild switch``` has to add or remove a container or change its address (due to changes in ```/etc/nixos/configuration.conf```)? Will the first option work?

**Comment by fpletz** [comment]
Yes. The first option is only problematic if you use the NixOS networking config to enslave interfaces to a bridge like:

    networking.bridges.br0.interfaces = [ "eth0" ];

If you would remove eth0 now, it will still be enslaved to the bridge after a `nixos-rebuild switch` until reboot or manual intervention.

**Comment by fpletz** [comment]
cc @kampfschlaefer

**Comment by volth** [comment]
The third option would be to scan ```/etc/containers/*.conf``` for strings like ```HOST_BRIDGE=br0```.
Although ```extraVeths``` would be lost.

**Comment by volth** [comment]
Here the patch which fixes the issue.

What is weird is although ```network-addresses-br0.service``` depends on ```br0-netdev.service``` transitively via ```sys-subsystem-net-devices-br0.device```, without explicit direct dependency the scripts try to start concurrently resulting in ```br0``` has no assigned IP after ```nixos-rebuild``` (it happens even without containers involved).

**Comment by volth** [comment]
Probably related https://github.com/NixOS/nixpkgs/issues/11176

**Comment by kampfschlaefer** [comment]
Hello (here and at #22547),

my experimentation has let me to some new knowledge:
* Its possible to just reload the bridge with a reload script in the bridge service. When only the list of (physical) interfaces enslaved to the bridge changes, this works nicely.
* When the rstp state changes from false to true, systemd tries to reload the bridge, but that fails as the userspace daemon is not there yet. And as the bridge service now depends on the inactive mstp.service, it is then stopped and waits for mstp to be up. Then it is started again, but of course now the containers have lost their connection.

For me this would be a workable solution so far as this keeps the containers connected by preventing bridge restart as good as possible.

Regarding my comment with the dependencies: I think these are not needed. At least not as "want", as far as I remember this would restart the bridge anytime the address changes?

**Comment by kampfschlaefer** [comment]
my idea for the fix is at #22850
